src/dmd/staticcond.d

```diff
@@ -415,9 +444,48 @@ private uint visualizeShort(Expression original, Expression instantiated,
         if (stack[i].inverted)
             buf.writeByte('!');
         buf.writestring(stack[i].orig.toChars);
+        if (stack[i].failMessage.length)
+        {
+            buf.writestring(": ");
+            buf.writestring(stack[i].failMessage);
+        }
         // here with no trailing newline
         if (i + 1 < stack.length)
             buf.writenl();
     }
     return cast(uint)stack.length;
 }
+
+/// Returns true if `exp` is a string or a struct with a toString method defined.
+bool isUserErrorStruct(Scope* sc, Expression e)
+{
+    import dmd.aliasthis;
+    import dmd.id;
+    import dmd.opover;
+
+    // code mostly from expressionsem.d callToString, but not emitting errors
```

Comment:
maybe it's possible to call code without emitting errors too? I don't know how but if that's the case then this function wouldn't be needed and that error silencing could be put around the struct to string code. (though I think having a test function for this has benefits as well)